libs/core/execution_base/include/hpx/execution_base/any_sender.hpp

```diff
@@ -0,0 +1,746 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/assert.hpp>
+#include <hpx/errors/error.hpp>
+#include <hpx/errors/throw_exception.hpp>
+#include <hpx/execution_base/sender.hpp>
+
+#include <cstddef>
+#include <cstring>
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx::detail {
+    template <typename T>
+    struct empty_vtable
+    {
+        static_assert(
+            sizeof(T) == 0, "No empty vtable type defined for given type T");
+    };
+
+    template <typename T>
+    struct get_empty_vtable
+    {
+        using empty_vtable_type = typename empty_vtable<T>::type;
+        static_assert(std::is_base_of_v<T, empty_vtable_type>,
+            "Given empty vtable type should be a base of T");
+
+        static T* call()
+        {
+            static empty_vtable_type empty;
+            return &empty;
+        }
+    };
+
+    template <typename Base, std::size_t EmbeddedStorageSize,
+        std::size_t AlignmentSize = sizeof(void*)>
+    class sbo_storage
+    {
+    protected:
+        using base_type = Base;
+        static constexpr std::size_t embedded_storage_size =
+            EmbeddedStorageSize;
+        static constexpr std::size_t alignment_size = AlignmentSize;
+
+        // The union has three members:
+        // - alignment_storage: This is only used to guarantee the alignment
+        //   given by alignment_size for the embedded storage.
+        // - heap_storage: A pointer to base_type that is used when objects
+        //   don't fit in the embedded storage.
+        // - embedded_storage: Fixed size array used for types that are at
+        //   most embedded_storage_size bytes large, and require at most
+        //   alignment_size alignment.
+        union
+        {
+            std::aligned_storage_t<embedded_storage_size, alignment_size>
+                alignment_storage;
+            base_type* heap_storage;
+            unsigned char embedded_storage[embedded_storage_size];
```

Comment:
Wouldn't it be sufficient to have the `alignment_storage` and the `heap_storage` members only?

Reply:
Yeah, it would. I mostly separated it because I thought it was clearer like this, at least when I wrote this initially. Now I'm not sure. I could go either way. Do you have a preference?

Reply:
I've removed the separate char array.